### ISSUE TITLE
How to make filename clickable in log output

### ISSUE THREAD

**Issue by eitel13** [Delgan/loguru#1012]
Hey @Delgan ! 👋 🙂 

I have a question, similar to this one: https://github.com/Delgan/loguru/issues/106

I'd like to make the path to the source file clickable in the log console. AFAIK, the way to do this for IntelliJ IDE's is to use a plugin called [awesome-console](https://github.com/anthraxx/intellij-awesome-console).

Following [this](https://github.com/Delgan/loguru/issues/106#issuecomment-594867003) solution from the above mentioned question, it's almost 🤏 what I need, but not quite right.

We have this code in an `__init__.py` file:

```
# set default logger log level to be able to intercept it
logging.basicConfig(
    handlers=[InterceptHandler()],
    level="DEBUG",
    force=True,
)

# reduce duplicate logs
# remove any loggers that may be associated with Loguru's logger
logger.remove()

# set log files directory path
log_directory_path = pathlib.Path(__file__).resolve().parent.parent / "logs"
test_run_time = time.strftime("%Y%m%d-%H%M%S", time.localtime())
log_file_name = f"testrun_{test_run_time}"

# set location
log_file_location = pathlib.Path(
    log_directory_path,
    log_file_name,
)

logger.disable("urllib3")
logger.add(
    sink=f"{log_file_location}.log",
    level="DEBUG",
    encoding="utf-8",
)
logger.add(
    sink=sys.stderr,
    level="INFO",
)
```

And this helper function (also grabbed I think from another question):

```
class InterceptHandler(logging.Handler):
    def emit(
        self,
        record: logging.LogRecord,
    ) -> None:
        # Get corresponding Loguru level if it exists.
        level: str | int

        try:
            level = logger.level(record.levelname).name
        except ValueError:
            level = record.levelno

        # Find caller from where originated the logged message.
        frame, depth = inspect.currentframe(), 0

        while frame and (depth == 0 or frame.f_code.co_filename == logging.__file__):
            frame = frame.f_back
            depth += 1

        logger.opt(
            depth=depth,
            exception=record.exc_info,
        ).log(
            level,
            record.getMessage(),
        )
```

As you can see, we don't mess with the logger format in this current setup, and the solution provided in the linked question above, _does_.

What this then does, is makes all the nice-looking logs from Loguru, super ugly.

# Updated Code

```
def custom_format(record):
    path = os.path.abspath(record["file"].path)
    record["extra"]["abspath"] = path
    return "{name} - {level} - {extra[abspath]} - {message}\n{exception}"


logger.add(
    sink=sys.stderr,
    level="INFO",
    format=custom_format,
)
```

# Before
![image](https://github.com/Delgan/loguru/assets/38290274/79e8a461-8ced-4301-8d9c-33b2acdeb95b)

# After
![image](https://github.com/Delgan/loguru/assets/38290274/07d79295-1a25-42a1-a6e1-cb0bd5b4c445)

Could you perhaps explain how to apply the formatter in a way that basically sticks to the default logger format but just changes the path to the file?

I'd really like to keep to the before style & format, with the only exception that the file name (`backend.utils.log_helpers`) be the clickable path to the actual source file.

**Comment by Delgan** [comment]
Hi @eitel13. ;)

I believe if you're using a recent version of Python, the absolute file path will be displayed when using `"{file.path}"` in your `format` (see [docs](https://loguru.readthedocs.io/en/stable/api/logger.html#record)), and you won't even need a custom formatting function:

```python
import sys
from loguru import logger

logger.remove()
logger.add(sys.stderr, format="{name} - {level} - {file.path} - {message}")
```

However, the `format` here is given as a very basic example and indeed does not contain as much information and colors than the default one. The default format used by Loguru is defined here: https://github.com/Delgan/loguru/blob/db0c8480b45ab8d79f7ff3eb554e91f8a4d08abf/loguru/_defaults.py#L31-L37

So basically, you need to adapt it to your preferences, for example:

```python
import sys
from loguru import logger

logging_format = (
    "<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | "
    "<level>{level: <8}</level> | "
    "<cyan>{file.path}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> - <level>{message}</level>"
)

logger.remove()
logger.add(sys.stderr, format=logging_format)
```

Note you can also configure your favorite format [as an environment variable](https://loguru.readthedocs.io/en/stable/api/logger.html#env) so that it becomes the default one.

**Comment by eitel13** [comment]
> Note you can also configure your favorite format [as an environment variable](https://loguru.readthedocs.io/en/stable/api/logger.html#env) so that it becomes the default one.
Man, this is great! 😮 

Thank you so much @Delgan 💪

I love how simple it is to _just use_ Loguru 🎉

**Comment by Delgan** [comment]
Great, glad you like Loguru, thanks you!

I'm closing this ticket then, since your issue seems solved. :+1:

**Comment by 5j9** [comment]
I'd love to see OSC 8 terminal hyperlinks in Loguru. That is `file_name:col` as text and full path (with `file:///` protocol) as the link. It could be the default or maybe as an optional feature. [`rich.logging.RichHandler`](https://github.com/Textualize/rich/) has this feature on by default and it looks awesome IMHO. (it has [a minor bug](https://github.com/Textualize/rich/issues/4093) though, but can be fixed easily).

The general idea is pretty simple. For a quick demo save the following script in a file and run it in vscode's terminal:
```python
from pathlib import Path


def link(file_path: Path, line_no: int, color: str):
    reset = '\x1b[0m'
    text = f'{file_path.name}:{line_no}'
    # Note: replace("\\", "/") for Windows paths, as_uri() does it already.
    uri = f'{file_path.as_uri()}#{line_no}'
    print(f'{color}\x1b]8;;{uri}\x1b\\{text}\x1b]8;;\x1b\\{reset}')


link(Path(__file__), 5, '\x1b[33m')  # Yellow link
```

Many of the popular terminals support this feature already, see:
https://github.com/Alhadis/OSC8-Adoption/
